process MEDAKA_CONSENSUS {
    tag "$sample_id"
    
    input:
    tuple val(sample_id), val(platform), path(bam), path(ref), val(strain_id)

    output:
    tuple val(sample_id), path("medaka_out/consensus.fasta"), val(strain_id)

    script:
    def model = 'r1041_e82_400bps_sup_variant_v4.1.0'
    
    """
    set -euo pipefail

    # Run Medaka Consensus
    medaka_consensus -i "$bam" -d "$ref" -o medaka_out -m "$model" -t ${task.cpus}

    # Create medaka_out directory in the current working directory
    mkdir -p ${task.workDir}/medaka_out

    # Copy the consensus.fasta from medaka_out to the new medaka_out directory
    cp medaka_out/consensus.fasta ${task.workDir}/medaka_out/consensus.fasta

    # Check if calls.vcf exists before copying it
    if [ -f medaka_out/calls.vcf ]; then
        cp medaka_out/calls.vcf ${task.workDir}/medaka_out/calls.vcf
        bgzip -c ${task.workDir}/medaka_out/calls.vcf > ${task.workDir}/medaka_out/calls.vcf.gz
        tabix -p vcf ${task.workDir}/medaka_out/calls.vcf.gz
    else
        echo "No VCF file generated by Medaka."
    fi
    """
}
